internal/controllers/application.go

```diff
@@ -0,0 +1,392 @@
+	}
+	return nil
+}
+
+/*
+ * Set application defaults. For existing applications this shouldn't do anything
+ */
+func (r *ApplicationReconciler) setApplicationDefaults(application *skiperatorv1alpha1.Application, ctx context.Context) {
+	application.FillDefaultsSpec()
+	if !ctrlutil.ContainsFinalizer(application, applicationFinalizer) {
+		ctrlutil.AddFinalizer(application, applicationFinalizer)
+	}
+
+	// Add labels to application
+	//TODO can we skip a step here?
+	if application.Labels == nil {
+		application.Labels = make(map[string]string)
+	}
+	maps.Copy(application.Labels, application.GetDefaultLabels())
+	maps.Copy(application.Labels, application.Spec.Labels)
+
+	// Add team label
+	if len(application.Spec.Team) == 0 {
+		if name, err := r.teamNameForNamespace(ctx, application); err == nil {
+			application.Spec.Team = name
+		}
+	}
+
+	application.FillDefaultsStatus()
+}
+
+func (r *ApplicationReconciler) isClusterReady(ctx context.Context) bool {
+	if !r.isCrdPresent(ctx, "servicemonitors.monitoring.coreos.com") {
+		return false
+	}
+	return true
+}
+
+func (r *ApplicationReconciler) teamNameForNamespace(ctx context.Context, app *skiperatorv1alpha1.Application) (string, error) {
+	ns := &corev1.Namespace{}
+	if err := r.GetClient().Get(ctx, types.NamespacedName{Name: app.Namespace}, ns); err != nil {
+		return "", err
+	}
+
+	teamValue := ns.Labels["team"]
+	if len(teamValue) > 0 {
+		return teamValue, nil
+	}
+	return "", fmt.Errorf("missing value for team label")
+}
+
+// Name in the form of "servicemonitors.monitoring.coreos.com".
+func (r *ApplicationReconciler) isCrdPresent(ctx context.Context, name string) bool {
+	result, err := r.GetApiExtensionsClient().ApiextensionsV1().CustomResourceDefinitions().Get(ctx, name, metav1.GetOptions{})
+	if err != nil || result == nil {
+		return false
+	}
+
+	return true
+}
+
+func handleApplicationCertRequest(_ context.Context, obj client.Object) []reconcile.Request {
+	cert, ok := obj.(*certmanagerv1.Certificate)
+	if !ok {
+		return nil
+	}
+
+	isSkiperatorOwned := cert.Labels["app.kubernetes.io/managed-by"] == "skiperator" &&
+		cert.Labels["skiperator.kartverket.no/controller"] == "application"
+
+	requests := make([]reconcile.Request, 0)
+
+	if isSkiperatorOwned {
+		requests = append(requests, reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Namespace: cert.Labels["application.skiperator.no/app-namespace"],
+				Name:      cert.Labels["application.skiperator.no/app-name"],
+			},
+		})
+	}
+
+	return requests
+}
+
+func isIngressGateway(gateway *networkingv1beta1.Gateway) bool {
+	match, _ := regexp.MatchString("^.*-ingress-.*$", gateway.Name)
+
+	return match
+}
+
+// TODO should be handled better
+func validateIngresses(application *skiperatorv1alpha1.Application) error {
+	var err error
+	hosts, err := application.Spec.Hosts()
+	if err != nil {
+		return err
+	}
+
+	// TODO: Remove/rewrite?
```

Comment:
Don't think we can do a regex match because `Ingresses` is JSON. I'd like to keep it I think, because removing it can cause unintended results when we try to generate resources (its used a lot of places). We should probably create a task to do this a better way though.

Reply:
I agree but it might be out of scope for this PR alone. It would be better to fix it in a separate, smaller PR and get more eyes on it.

Reply:
Think I will keep the TODO, and we can revisit for a proper solution. Probably better if I focus on getting dynamic port allocation working.